src/CanvasRenderingContext2d.cc

```diff
@@ -662,10 +663,13 @@ NAN_METHOD(Context2d::PutImageData) {
         uint8_t b = *srcRow++;
         uint8_t a = *srcRow++;
 
-        // argb
-        // performance optimization: fully transparent/opaque pixels can be
-        // processed more efficiently.
-        if (a == 0) {
+        if (dstRow < dstStart) {
```

Comment:
It would be nice to avoid more branches inside the innermost loop. I'd look at modifying the `rows`, `cols`, `dst` and `src` values around lines 615-616 instead. Lines 624-639 should provide good guidance; they handle negative values for the 7-argument form.